### PR TITLE
Feature/OP-1228: Limit Concurrent User Sessions

### PR DESCRIPTION
**Pull request by joelbcastillo**
This PR implements a limit to concurrent user sessions. Users are alerted when trying to login to a new computer before having logged out of their old session. 

Session IDs are stored in the Users table in the database. If the value is a valid session ID, then the user is notified that they are being logged out of all other sessions and the value in the Users table is updated to be the current session.

Note, if the user opens the old session, they will be able to invalidate the "new" session by clicking login. This is due to the limitation of using OAuth which requires us to send the user to the IdP logout page in order to invalidate the IdP session on the computer. This is not something we can currently force the user to do in this PR. 

If anyone can figure out a way to handle this to avoid this, please let me know.

### CODE REVIEW

**Comment by johnyu95** [comment]
I found a few ways to bypass this feature but I'm not sure if they're intended behavior:

Method 1
- Have 2 windows of chrome open
- Login to the first window
- Login to the second window. When you see the concurrent session modal, change the url to get rid of the '?fresh_login=True' and press enter

You now have 2 sessions logged in. Both sessions can navigate the site without the concurrent session modal appearing again. I confirmed it was running concurrently by going to the manage account page and changing some information. If you refresh the other window the changes from the first window will appear.

Method 2
- Have 2 windows of chrome open
- Login to the first window
- Login to the second window and press continue on the concurrent session modal
- On the first window, click "Manage Account" 

This will cause the concurrent session modal to appear on the first window and effectively keep you logged in when you supposedly logged out using the modal in the second window.

File: app/auth/utils.py [hunk 1]
```diff
@@ -54,6 +60,24 @@ def user_loader(user_id):
     return Users.query.filter_by(guid=user_id[0], auth_user_type=user_id[1]).first()
 
 
+def destroy_session(user_id):
```
Comment (johnyu95) [inline app/auth/utils.py]: This function is never called when logging out. I think you use "session.destroy()" instead.

File: app/auth/views.py [hunk 2]
```diff
@@ -209,9 +219,40 @@ def ldap_login():
 
 
 @auth.route('/ldap_logout', methods=['GET'])
-def ldap_logout(timed_out=None):
+def ldap_logout(timed_out=False, forced_logout=False):
+    if forced_logout:
```
Comment (johnyu95) [inline app/auth/views.py]: You can take this if statement out.

File: app/auth/views.py [hunk 3]
```diff
@@ -209,9 +219,40 @@ def ldap_login():
...
+    if timed_out:
         flash("Your session timed out. Please login again", category='info')
     return redirect(url_for('main.index'))
+
+
+@auth.route('/oauth_logout', methods=['GET'])
+def oauth_logout():
+    timed_out = eval_request_bool(request.args.get('timeout'))
+    forced_logout = eval_request_bool(request.args.get('forced_logout'))
+    print("Timed Out: {}, Forced Logout: {}".format(timed_out, forced_logout))
```
Comment (johnyu95) [inline app/auth/views.py]: Are these print statements just for development debugging?

**Comment by joelbcastillo** [comment]
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[app/auth/utils.py, line 63 at r3](https://reviewable.io:443/reviews/cityofnewyork/nycopenrecords/229#-KwGx-kWLxVYGSWjOyVQ-r3-63:-KwH1J95o67xMfJUyCrh:b-gt4agc) ([raw file](https://github.com/cityofnewyork/nycopenrecords/blob/869260472c37c4c22cd1f6014235308e66814b31/app/auth/utils.py#L63)):*
<details><summary><i>Previously, johnyu95 wrote…</i></summary><blockquote>

This function is never called when logging out. I think you use "session.destroy()" instead.
</blockquote></details>

The function was left over from before. It doesn't work as intended. Removing it now.

---

*[app/auth/views.py, line 223 at r3](https://reviewable.io:443/reviews/cityofnewyork/nycopenrecords/229#-KwGx-kWLxVYGSWjOyVR-r3-223:-KwH1cS1p19V6K_s-2bm:b-v6s1o3) ([raw file](https://github.com/cityofnewyork/nycopenrecords/blob/869260472c37c4c22cd1f6014235308e66814b31/app/auth/views.py#L223)):*
<details><summary><i>Previously, johnyu95 wrote…</i></summary><blockquote>

You can take this if statement out.
</blockquote></details>

Removed. 

---

*[app/auth/views.py, line 236 at r3](https://reviewable.io:443/reviews/cityofnewyork/nycopenrecords/229#-KwGx-kWLxVYGSWjOyVR-r3-236:-KwH1R8w7Mi5wdLxxrpL:bgob75y) ([raw file](https://github.com/cityofnewyork/nycopenrecords/blob/869260472c37c4c22cd1f6014235308e66814b31/app/auth/views.py#L236)):*
<details><summary><i>Previously, johnyu95 wrote…</i></summary><blockquote>

Are these print statements just for development debugging?
</blockquote></details>

Yep, removing them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cityofnewyork/nycopenrecords/229)*

**Comment by joelbcastillo** [comment]
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[app/auth/views.py, line 236 at r3](https://reviewable.io:443/reviews/cityofnewyork/nycopenrecords/229#-KwGx-kWLxVYGSWjOyVR-r3-236:-KwH20DejvksBBfkW81J:b-896fix) ([raw file](https://github.com/cityofnewyork/nycopenrecords/blob/869260472c37c4c22cd1f6014235308e66814b31/app/auth/views.py#L236)):*
<details><summary><i>Previously, joelbcastillo (Joel Castillo) wrote…</i></summary><blockquote>

Yep, removing them.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cityofnewyork/nycopenrecords/229)*

**Comment by joelbcastillo** [comment]
Review status: 0 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[app/auth/views.py, line 223 at r3](https://reviewable.io:443/reviews/cityofnewyork/nycopenrecords/229#-KwGx-kWLxVYGSWjOyVR-r3-223:-KwH22VGUTvvXhVB9ZPw:b-896fix) ([raw file](https://github.com/cityofnewyork/nycopenrecords/blob/869260472c37c4c22cd1f6014235308e66814b31/app/auth/views.py#L223)):*
<details><summary><i>Previously, joelbcastillo (Joel Castillo) wrote…</i></summary><blockquote>

Removed. 
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cityofnewyork/nycopenrecords/229)*

**Comment by joelbcastillo** [comment]
@johnyu95 Any update on this?

Thanks

**Comment by johnyu95** [comment]
Looks good.

**Pull request merged**